src/rpp/rpp/schedulers/trampoline_scheduler.hpp

```diff
@@ -0,0 +1,101 @@
+//
+//          Copyright Aleksey Loginov 2022 - present.
+//                            TC Wang 2022 - present.
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          https://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/victimsnino/ReactivePlusPlus
+//
+
+#pragma once
+
+#include <rpp/schedulers/fwd.hpp>                       // own forwarding
+#include <rpp/schedulers/details/worker.hpp>            // worker
+#include <rpp/subscriptions/composite_subscription.hpp> // lifetime
+#include <rpp/schedulers/details/queue_worker_state.hpp>// state
+
+#include <concepts>
+#include <chrono>
+#include <functional>
+#include <thread>
+
+thread_local bool queue_in_use = false;
+thread_local rpp::schedulers::details::queue_worker_state queue{};
+
+namespace rpp::schedulers
+{
+
+template <typename F>
+struct scope_guard {
+    explicit scope_guard(F&& f) : m_f(std::move<F>(std::forward<F>(f))) {}
+    ~scope_guard() noexcept { m_f(); }
+    F m_f;
+};
+
+/**
+ * \brief schedules execution of schedulables via queueing tasks to the caller thread with priority to time_point and order
+ *
+ * \par Example
+ * \snippet trampoline.cpp trampoline
+ *
+ * \ingroup schedulers
+ */
+class trampoline final : public details::scheduler_tag
+{
+public:
+    class worker_strategy
+    {
+    public:
+        explicit worker_strategy(const rpp::composite_subscription& subscription)
+            : m_sub{subscription}, m_stop{std::make_shared<std::stop_source>()} {};
+
+        void defer_at(time_point time_point, std::invocable auto&& fn) const
+        {
+            if (!m_sub.is_subscribed())
+                return;
+
+            queue.emplace(time_point, std::forward<decltype(fn)>(fn));
+            if (queue_in_use)
+                return;
+
+            queue_in_use = true;
+            scope_guard cleanup([&]() noexcept
+            {
+                // If error occurs, we still want the thread-local state to be reset.
+                queue_in_use = false;
+                if (!m_sub.is_subscribed())
+                    queue.reset();
+            });
+
+            auto stop_token = m_stop->get_token();
+            m_sub.add([stop = m_stop]()
+            {
+                // The following blocking-pop exits when subscription is unsubscribed in other thread.
+                stop->request_stop();
+            });
+            while (!queue.is_empty() && m_sub.is_subscribed())
+            {
+                std::function<void()> schedulable{};
+                if (queue.pop_with_wait(schedulable, stop_token))
+                {
+                    schedulable();
+                    schedulable = {};
+                }
+            }
+        }
+
+        static time_point now() { return clock_type::now(); }
+
+    private:
+        rpp::composite_subscription m_sub;
+        std::shared_ptr<std::stop_source> m_stop;
```

Comment:
you don't need it as shared_ptr. You can create it during `defer_at` function. Add callback to subscription and remove/unsubscribe it after while loop to remove local variable from m_sub state

Reply:
I used a shared pointer for shared ownerships. Do you mean something like below in `defer_at()`?

```
std::stop_source stop;
auto stop_token = stop.get_token();

m_sub.add([stop = std::move(stop)] { // move stop since subscription callback is the sole owner.
  stop.request_stop();
});

while (...) {
  if (queue.pop_wait(..., stop_token) ...
}
```

Reply:
I feel the `stop_source` has the same lifetime to `m_sub` and thus I lift it as a class member.